// @atlaskit/theme/dist/cjs/elevation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.e500 = exports.e400 = exports.e300 = exports.e200 = exports.e100 = undefined;

var _colors = require('./colors');

var colors = _interopRequireWildcard(_colors);

var _themed = require('./utils/themed');

var _themed2 = _interopRequireDefault(_themed);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

// Cards on a board
var e100 = exports.e100 = (0, _themed2.default)({
  light: 'box-shadow: 0 1px 1px ' + colors.N50A + ', 0 0 1px 0 ' + colors.N60A + ';',
  dark: 'box-shadow: 0 1px 1px ' + colors.DN50A + ', 0 0 1px ' + colors.DN60A + ';'
});

// Inline dialogs
var e200 = exports.e200 = (0, _themed2.default)({
  light: 'box-shadow: 0 4px 8px -2px ' + colors.N50A + ', 0 0 1px ' + colors.N60A + ';',
  dark: 'box-shadow: 0 4px 8px -2px ' + colors.DN50A + ', 0 0 1px ' + colors.DN60A + ';'
});

// Modals
var e300 = exports.e300 = (0, _themed2.default)({
  light: 'box-shadow: 0 8px 16px -4px ' + colors.N50A + ', 0 0 1px ' + colors.N60A + ';',
  dark: 'box-shadow: 0 8px 16px -4px ' + colors.DN50A + ', 0 0 1px ' + colors.DN60A + ';'
});

// Panels
var e400 = exports.e400 = (0, _themed2.default)({
  light: 'box-shadow: 0 12px 24px -6px ' + colors.N50A + ', 0 0 1px ' + colors.N60A + ';',
  dark: 'box-shadow: 0 12px 24px -6px ' + colors.DN50A + ', 0 0 1px ' + colors.DN60A + ';'
});

// Flag messages (notifications)
var e500 = exports.e500 = (0, _themed2.default)({
  light: 'box-shadow: 0 20px 32px -8px ' + colors.N50A + ', 0 0 1px ' + colors.N60A + ';',
  dark: 'box-shadow: 0 20px 32px -8px ' + colors.DN50A + ', 0 0 1px ' + colors.DN60A + ';'
});
